### PR TITLE
Replaced param with input val channel

### LINKED ISSUE #594: [BUG] `fastp` module includes `params` rather than input val
## Check Documentation

I have checked the following places for your error:

- [x] [nf-core website: troubleshooting](https://nf-co.re/usage/troubleshooting)
- [x] [nf-core/module documentation](https://github.com/nf-core/modules/blob/master/README.md)

## Description of the bug

`fastp` currently is not following best practice/guidelines as it includes an internal `params` value:

https://github.com/nf-core/modules/blob/0b40798d1b59a28bd86aa5b558a0793614be1efa/modules/fastp/main.nf#L37

This should be replaced with a `val` input declaration

## Steps to reproduce

Steps to reproduce the behaviour:

NA

## Expected behaviour

Follows current guidelines

## Log files

Have you provided the following extra information/files:

- [ ] The command used to run the module
- [ ] The `.nextflow.log` file 

## System

NA

## Additional context

Slack convo: https://nfcore.slack.com/archives/CJRH30T6V/p1626775872243800

### PR DESCRIPTION
**Pull request by jfy133**
## PR checklist

Current `fastp` module includes a `params` option which isn't allowed according to the nf-core module guidelines. This replaces the `params.save_trimmed_fail` with a input `value` channel

Note: I'm not sure if this is the most optimal implementation here, as it means it adds a _required_ input channel , and I'm not 100% of this parameter can be considered to be 'required as it requires the explicit creation of an extra channel for a parameter maybe irrelevant for some cases of `fastp`.

Closes #594 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/fastp/main.nf [hunk 1]
```diff
@@ -20,6 +20,7 @@ process FASTP {
 
     input:
     tuple val(meta), path(reads)
+    val save_trimmed_fail
```
Comment (JoseEspinosa) [inline modules/fastp/main.nf:23]:
```suggestion
    val   save_trimmed_fail
```

File: tests/modules/fastp/main.nf [hunk 2]
```diff
@@ -11,8 +11,9 @@ workflow test_fastp_single_end {
     input = [ [ id:'test', single_end:true ], // meta map
               [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
             ]
+    save_trimmed_fail = Channel.value(false)
```
Comment (JoseEspinosa) [inline tests/modules/fastp/main.nf:14]: Didn't know this one! :love_you_gesture:

File: tests/modules/fastp/main.nf [hunk 3]
```diff
@@ -23,7 +24,33 @@ workflow test_fastp_paired_end {
...
 
-    FASTP ( input )
+    FASTP ( input, save_trimmed_fail )
+}
+
+//
+// Test with single-end data with saving trimming fails
+//
+workflow test_fastp_single_end_trimfail {
```
Comment (JoseEspinosa) [inline tests/modules/fastp/main.nf:35]: For consistency maybe:

```suggestion
workflow test_fastp_single_end_trim_fail {
```
```

File: tests/modules/fastp/main.nf [hunk 4]
```diff
@@ -23,7 +24,33 @@ workflow test_fastp_paired_end {
...
+            ]
+    save_trimmed_fail = Channel.value(true)
+
+    FASTP ( input, save_trimmed_fail )
+}
+
+//
+// Test with paired-end data with saving trimming fails
+//
+workflow test_fastp_paired_end_trimfail {
```
Comment (JoseEspinosa) [inline tests/modules/fastp/main.nf:47]: Same here:
```suggestion
workflow test_fastp_paired_end_trim_fail {
```

File: tests/modules/fastp/test.yml [hunk 5]
```diff
@@ -36,3 +36,48 @@
       md5sum: e2257263668dc8a75d95475099fb472d
     - path: output/fastp/test_2.trim.fastq.gz
       md5sum: 9eff7203596580cc5e42aceab4a469df
+
+- name: fastp test_fastp_single_end_trimfail
```
Comment (JoseEspinosa) [inline tests/modules/fastp/test.yml:40]:
```suggestion
- name: fastp test_fastp_single_end_trim_fail
```

File: tests/modules/fastp/test.yml [hunk 6]
```diff
@@ -36,3 +36,48 @@
...
+  command: nextflow run tests/modules/fastp -entry test_fastp_single_end_trimfail -c tests/config/nextflow.config
```
Comment (JoseEspinosa) [inline tests/modules/fastp/test.yml:41]:
```suggestion
command: nextflow run tests/modules/fastp -entry test_fastp_single_end_trim_fail -c tests/config/nextflow.config
```

File: tests/modules/fastp/test.yml [hunk 7]
```diff
@@ -36,3 +36,48 @@
...
+        - "Q20 bases: 12922(92.9841%)"
+        - "reads passed filter: 99"
+    - path: output/fastp/test.trim.fastq.gz
+      md5sum: e2257263668dc8a75d95475099fb472d
+    - path: output/fastp/test.fastp.json
+      md5sum: ee65a46d6e59fa556f112727b8a902ce
+    - path: output/fastp/test.fail.fastq.gz
+      md5sum: de315d397c994d8e66bafc7a8dc11070
+
+- name: fastp test_fastp_paired_end_trimfail
```
Comment (JoseEspinosa) [inline tests/modules/fastp/test.yml:60]:
```suggestion
- name: fastp test_fastp_paired_end_trim_fail
```

File: tests/modules/fastp/test.yml [hunk 8]
```diff
@@ -36,3 +36,48 @@
...
+  command: nextflow run tests/modules/fastp -entry test_fastp_paired_end_trimfail -c tests/config/nextflow.config
```
Comment (JoseEspinosa) [inline tests/modules/fastp/test.yml:61]:
```suggestion
  command: nextflow run tests/modules/fastp -entry test_fastp_paired_end_trim_fail -c tests/config/nextflow.config
```

**Review by JoseEspinosa — CHANGES_REQUESTED** [review]
LGTM! Just some small suggestions :rocket:

File: tests/modules/fastp/main.nf (hunk 2 above)
Comment (drpatelh) [reply tests/modules/fastp/main.nf:14]: `save_trimmed_fail = false` is fine :)

File: tests/modules/fastp/main.nf [hunk 9]
```diff
@@ -23,7 +24,33 @@ workflow test_fastp_paired_end {
               [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                 file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
             ]
+    save_trimmed_fail = Channel.value(false)
```
Comment (drpatelh) [inline tests/modules/fastp/main.nf:27]:
```suggestion
    save_trimmed_fail = false
```

File: tests/modules/fastp/main.nf [hunk 10]
```diff
@@ -23,7 +24,33 @@ workflow test_fastp_paired_end {
...
+
+//
+// Test with paired-end data with saving trimming fails
+//
+workflow test_fastp_paired_end_trimfail {
+    input = [ [ id:'test', single_end:false ], // meta map
+              [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+                file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+            ]
+    save_trimmed_fail = Channel.value(true)
```
Comment (drpatelh) [inline tests/modules/fastp/main.nf:52]:
```suggestion
    save_trimmed_fail = true
```

**Review by drpatelh — COMMENTED** [review]
Thanks @jfy133 ! Good to go once @JoseEspinosa comments have been addressed :) Unfortunately, this is the only way we can deal with optional inputs that require parameter evaluation until something better comes along in NF.

File: tests/modules/fastp/main.nf (hunk 2 above)
Comment (JoseEspinosa) [reply tests/modules/fastp/main.nf:14]: Even better :heart_eyes:

Comment (jfy133) [inline tests/modules/fastp/main.nf:14]:
```suggestion
    save_trimmed_fail = false
```

File: tests/modules/fastp/main.nf [hunk 11]
```diff
@@ -23,7 +24,32 @@ workflow test_fastp_paired_end {
...
 }
 
+//
+// Test with single-end data with saving trimming fails
+//
+workflow test_fastp_single_end_trim_fail {
+    input = [ [ id:'test', single_end:true ], // meta map
+              [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
+            ]
+    save_trimmed_fail = Channel.value(true)
```
Comment (jfy133) [inline tests/modules/fastp/main.nf:39]:
```suggestion
    save_trimmed_fail = true
```

**Comment by jfy133** [comment]
> Thanks @jfy133 ! Good to go once @JoseEspinosa comments have been addressed :) Unfortunately, this is the only way we can deal with optional inputs that require parameter evaluation until something better comes along in NF.

Given this is is not a 'backwards compatible' change, is there any mechanism to inform people that they would need to add a new input declaration at workflow level, when updating this module?

**Review by JoseEspinosa — APPROVED** [review]
Thanks @jfy133 ! Ready to go :rocket

**Comment by JoseEspinosa** [comment]
> > Thanks @jfy133 ! Good to go once @JoseEspinosa comments have been addressed :) Unfortunately, this is the only way we can deal with optional inputs that require parameter evaluation until something better comes along in NF.
> 
> Given this is is not a 'backwards compatible' change, is there any mechanism to inform people that they would need to add a new input declaration at workflow level, when updating this module?

I don't think there is such a thing but maybe @drpatelh have some thoughts on this :thinking:

**Comment by jfy133** [comment]
Thanks! Merging now!

> > > Thanks @jfy133 ! Good to go once @JoseEspinosa comments have been addressed :) Unfortunately, this is the only way we can deal with optional inputs that require parameter evaluation until something better comes along in NF.
> > 
> > 
> > Given this is is not a 'backwards compatible' change, is there any mechanism to inform people that they would need to add a new input declaration at workflow level, when updating this module?
> 
> I don't think there is such a thing but maybe @drpatelh have some thoughts on this

Discussion on slack and on #597 started :)

**Comment by JoseEspinosa** [comment]
Just saw it :smile: 
Too many "channels" of discussion

**Pull request merged**